static/posts/pattern-matching/post.md

```diff
@@ -0,0 +1,952 @@
+
+```c
+#include <math.h>
+
+// ...
+
+double shape_area(Shape shape) {
+  Dimensions ds = shape.dimensions;
+  if (shape.type == CIRCLE) {
+    return M_PI * ds.r * ds.r;
+  } else if (shape.type == RECTANGLE) {
+    return ds.w * ds.h;
+  } else {
+    double s = (ds.a + ds.b + ds.c) / 2;
+    return sqrt(s * (s - ds.a) * (s - ds.b) * (s - ds.c));
+  }
+}
+```
+
+We can also write a short program to test out our function:
+
+```c
+#include <stdio.h>
+
+// ...
+
+int main() {
+  size_t size = 3;
+  Shape shapes[size];
+  
+  // Define the shapes
+
+  // Area should be pi
+  shapes[0].type = CIRCLE;
+  shapes[0].dimensions.r = 1.0;
+
+  // Area should be 6.0
+  shapes[1].type = RECTANGLE;
+  shapes[1].dimensions.w = 2.0;
+  shapes[1].dimensions.h = 3.0;
+
+  // Area should be 6.0
+  shapes[2].type = TRIANGLE;
+  shapes[2].dimensions.a = 3.0;
+  shapes[2].dimensions.b = 4.0;
+  shapes[2].dimensions.c = 5.0;
+
+  double totalArea = 0;
+  for (size_t i = 0; i < size; i++) {
+    totalArea += shape_area(shapes[i]);
+  }
+
+  // Total area should be pi + 6.0 + 6.0 = 15.14...
+  printf("Total area: %f\n", totalArea);
+
+  return 0;
+}
+```
+
+When we compile and run our program, the output is as expected:
+
+```text
+$ gcc shapes.c -o shapes -lm
+$ ./shapes
+Total area: 15.141593
+```
+
+Now let's modify our types and array of shapes to include a new kind of shape:
+
+```c
+// ...
+
+typedef enum {
+  CIRCLE,
+  RECTANGLE,
+  triangle,
+  PARALLELOGRAM,
+} ShapeType;
+
+typedef union {
+  struct { double r; }; // a circle
+  struct { double w; double h; }; // a rectangle
+  struct { double a; double b; double c; }; // a triangle
+  struct { double x; double y; double t; }; // a parallelogram
+} Dimensions;
+
+// ...
+
+int main() {
+  size_t size = 3;
+  Shape shapes[size];
+  
+  // Define the shapes
+
+  // Area should be pi
+  shapes[0].type = CIRCLE;
+  shapes[0].dimensions.r = 1.0;
+
+  // Area should be 6.0
+  shapes[1].type = RECTANGLE;
+  shapes[1].dimensions.w = 2.0;
+  shapes[1].dimensions.h = 3.0;
+
+  // Area should be 6.0
+  shapes[2].type = triangle;
+  shapes[2].dimensions.a = 3.0;
+  shapes[2].dimensions.b = 4.0;
+  shapes[2].dimensions.c = 5.0;
+
+  // Area should be 6.0
+  shapes[3].type = PARALLELOGRAM;
+  shapes[3].dimensions.x = 2.0;
+  shapes[3].dimensions.y = 3.0;
+  shapes[3].dimensions.t = M_PI / 2;
+
+  double totalArea = 0;
+  for (size_t i = 0; i < size; i++) {
+    totalArea += shape_area(shapes[i]);
+  }
+
+  // Total area should be pi + 6.0 + 6.0 + 6.0 = 21.14...
+  printf("Total area: %f\n", totalArea);
+
+  return 0;
+}
+```
+
+We can compile and run our program:
+
+```text
+$ gcc shapes.c -o shapes -lm
+$ ./shapes
+Total area: 16.579133
+```
+
+Wait, that's not right... It should have printed 21.14... What's going on here?
+There weren't any compile errors, everything should be fine right?
+
+Wrong. The C compiler will compile this without complaint. It will even run the
+program for you without throwing an error or
```

Comment:
I don't understand how this sentence ends, but I'm assuming it will display properly and fine. :)

Reply:
Yep, that's just how you write a link in Markdown (what .md stands for)